File: spectre.ml
let measure_access_time password guess index =
  let start_time = Sys.time () in
  (* Speculative execution *)
  if String.get password index = guess then
    let _ =
      let temp = ref 0 in
      for _ = 1 to 1000 do
        temp := !temp + 1
      done
    in
    Sys.time () -. start_time
  else
    Sys.time () -. start_time

let crack_password password =
  let characters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~" in
  let cracked_password = Array.make (String.length password) '*' in
  for i = 0 to String.length password - 1 do
    let max_time = ref 0.0 in
    let max_char = ref '*' in
    String.iter (fun c ->
      let total_time = ref 0.0 in
      for _ = 1 to 100 do
        let access_time = measure_access_time password c i in
        total_time := !total_time +. access_time
      done;
      let avg_time = !total_time /. 100.0 in
      if avg_time > !max_time then (
        max_time := avg_time;
        max_char := c
      )
    ) characters;
    cracked_password.(i) <- !max_char
  done;
  String.init (Array.length cracked_password) (fun i -> cracked_password.(i))

let () =
  let password = "hTQU?K4asW.xl,T\\Gm?vj48`}" in
  Printf.printf "Original Password: %s\n" password;
  let cracked_password = crack_password password in
  Printf.printf "Cracked Password: %s\n" cracked_password

File: spectre.rs
use std::time::Instant;

fn measure_access_time(password: &str, guess: char, index: usize) -> f64 {
    let start_time = Instant::now();
    
    // Speculative execution
    if password.chars().nth(index) == Some(guess) {
        let mut _temp = 0;
        for _ in 0..1000 {
            _temp += 1;
        }
    }
    
    let elapsed_time = start_time.elapsed();
    elapsed_time.as_secs_f64()
}

fn crack_password(password: &str) -> String {
    let characters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~";
    let mut cracked_password = vec!['*'; password.len()];
    
    for (i, _) in password.chars().enumerate() {
        let mut max_time = 0.0;
        let mut max_char = '*';
        
        for c in characters.chars() {
            let mut total_time = 0.0;
            for _ in 0..100 {
                let access_time = measure_access_time(password, c, i);
                total_time += access_time;
            }
            let avg_time = total_time / 100.0;
            
            if avg_time > max_time {
                max_time = avg_time;
                max_char = c;
            }
        }
        
        cracked_password[i] = max_char;
    }
    
    cracked_password.iter().collect()
}

fn main() {
    let password = "hTQU?K4asW.xl,T\\Gm?vj48`}";
    println!("Original Password: {}", password);
    
    let cracked_password = crack_password(password);
    println!("Cracked Password: {}", cracked_password);
}
